edd-discount-widget.php

```diff
@@ -134,21 +141,69 @@ function form( $instance ) {
         }
 
 
+        /**
+         * Obtains the EDD discounts from the site URL.
+         * 
+         * @param mixed $site_url      The URL to extract the discounts from.
+         * @param mixed $api_key       The API key for the site.
+         * @param mixed $api_token     The API token for the site.
+         * @param int   $max_discounts The maximum number of discounts to obtain.
+         * 
+         * @return array|WP_Error      An array of discounts or a WP_Error object.
+         */
         function get_discounts( $site_url, $api_key, $api_token, $max_discounts ) {
+
+            $site_url      = sanitize_text_field( $site_url );
+            $api_key       = sanitize_text_field( $api_key );
+            $api_token     = sanitize_text_field( $api_token );
+            $max_discounts = absint( $max_discounts );
+
+            if ( empty( $site_url ) || empty( $api_key ) || empty( $api_token ) ) {
+                return new WP_Error(
+                    'edd-discounts-no-settings',
+                    __( 'Widget settings are incomplete.', 'edd-discounts-widget' )
+                );
+            }
+
             $options = array(
                 'timeout'   => 5
             );
 
             $temp_url = parse_url( $site_url );
-            if( !$temp_url['scheme'] == 'http' && !$temp_url['scheme'] == 'https' )
-                $site_url = 'http://' . $site_url;
+            if ( isset( $temp_url['scheme'] ) ) {
+                if( !$temp_url['scheme'] == 'http' && !$temp_url['scheme'] == 'https' ) {
```

Comment:
You mean something like this?

``` php
if ( is_array( $temp_url ) && ! array_key_exists( 'scheme', $temp_url ) ) {
    $site_url = 'http://' . $site_url;
}
```

It's purpose is to add 'http://' if there's no scheme isn't it?

Reply:
Actually that's what I'm assuming was the intention of the code. In practice it would add `http://` if you put any URI with any scheme besides `http://` and `http://`. With no scheme it generates an error, and with any other type of scheme the URI would be a mess.

Reply:
I think what @ashleyfae means is that instead of 
`if( !$temp_url['scheme'] == 'http' && !$temp_url['scheme'] == 'https' )` 

you can shorten it down to 

`if ( ! in_array( 'http' || 'https', $temp_url ) )` and then keep the body of the function the same.

Reply:
Unfortunately `in_array()` doesn't work with associative arrays, which is what `parse_url()` returns (assuming a URL style string is passed to it).

Reply:
`parse_url()` results.

With scheme included in URL:

![image](https://user-images.githubusercontent.com/16655270/129333923-347073d5-441f-4b8e-ad33-54939571b2b4.png)

Without a scheme:

![image](https://user-images.githubusercontent.com/16655270/129334009-a37d2d54-ff2d-4e91-b6e8-448cd01cdaee.png)

Reply:
I think the purpose of the original code was to add `http://` to the beginning of the URL if it didn't have a scheme.

Reply:
You could try something like:

```php
			if ( ! empty( $temp_url['scheme'] ) && ! in_array( $temp_url['scheme'], array( 'http', 'https' ), true ) ) {
				$site_url = 'http://' . $site_url;
			}
```

...that would allow you to combine the check for the scheme being set with the protocol.

Reply:
What @robincornett suggested above is what I was getting at. Though in this instance I'd go with `isset( $temp_url['scheme'] )` instead of an `! empty()` check.

Reply:
Sorted :)